In [68]:
from community_detection import *      
from time import *

In [69]:
tab_amis = lecture_reseau("files/Communaute0.csv")
print(tab_amis)

['Alice', 'Bob', 'Bob', 'Charlie', 'Alice', 'Dominique', 'Bob', 'Dominique']


In [70]:
# Temps d'execution de la fonction dico_reseau

def temps_dico_reseau(tableau):
    tic = time()
    dico_amis = dico_reseau(tableau)
    tac = time()
    temps = round( 1000 * (tac-tic) , 3)
    
    return  temps

print (temps_dico_reseau(tab_amis))    


# Moyenne de temps d'execution de dico_reseau sur n test

def temps_dico_reseau_moyen(tableau_amis,n):
    i = 0
    tab = []
    somme = 0
    moyenne = 0
    
    while i < n:
        tab.append(temps_dico_reseau(tableau_amis))
        i += 1

    for i in range(0 , len(tab)):
        somme += tab[i]
    
    moyenne = somme/len(tab)
    return moyenne
    

temps_dico_reseau_moyen(tab_amis,1000)



0.0


0.005983

In [87]:
# Temps d'execution de la fonction create_network

def temps_create_network(tableau):
    tic = time()
    network = create_network(tableau)
    tac = time()
    
    temps = round( 1000 * (tac-tic) , 3)

    return temps



# Moyenne de temps d'execution de create_network sur n test

def temps_create_network_moyens(tableau_amis,n):
    i = 0
    tab = []
    somme = 0
    moyenne = 0
    
    while i < n:
        tab.append(temps_create_network(tableau_amis))
        i += 1

    for j in range(0 , len(tab)):
        somme += tab[j]
    
    moyenne = somme/len(tab)

    return moyenne


temps_create_network_moyens(tab_amis,1000)

0.001994

En moyenne (meme si les resultats changent legerement a chaques lancé), la fonction dico_reseau() de la premiere partie met
entre 0,020 et 0,030 secondes a s'executer ce qui est majoritairement plus long que la fonction create_network() qui elle met entre 0,001 et 0,005 seconde a s'executer.
Cela est dut au fait que la fonction dico_reseau appel une autre fonction dans une boucle while ce qui fait monter sa complexité a O(n²) alors que la fonction create_network() ne parcours la liste qu'une fois et ne depend que d'elle meme ce qui la rend plus rapide. La complexité de temps theorique de create_network est O(n)

In [91]:

# Temps d'execution de la fonction find_community_by_decreasing_popularity 
network = dico_reseau('files/Communaute0.csv')

def temps_find_community_by_decreasing_popularity(network):
    tic = time()
    test = find_community_by_decreasing_popularity(network)
    tac = time()
    temps = round( 1000 * (tac-tic) , 3)
    
    return  temps


# Moyenne de temps d'execution de find_community_by_decreasing_popularity sur n test

def moyenne_temps_find_community_by_decreasing_popularity(network,n):
    tab = []
    i = 0
    moyenne = 0
    somme = 0
    while i < n:
        tab.append(temps_find_community_by_decreasing_popularity(network))
        i+=1
    
    for j in range(0 , len(tab)):
        somme += tab[j]
    
    moyenne = somme/len(tab)

    return moyenne

moyenne_temps_find_community_by_decreasing_popularity(network,1000)

0.12124400000000003

In [93]:

# Temps d'execution de la fonction find_community_from_person
network = dico_reseau('files/Communaute0.csv')

def temps_find_community_from_person(network):
    tic = time()
    person = les_plus_pop(network)
    person = person[0]
    tac = time()

    tic2 = time()
    test = find_community_from_person(network, person)
    tac2 = time()

    temps = round( 1000 * ((tac-tic)+(tac2-tic2)) , 3)

    return temps



# Moyenne de temps d'execution de find_community_from_person sur n test

def moyenne_temps_find_community_from_person(network,n):
    tab = []
    i = 0
    moyenne = 0
    somme = 0

    while i < n:
        tab.append(temps_find_community_from_person(network))
        i+=1
    
    for j in range(0, len(tab)):
        somme += tab[j]

    moyenne = somme/len(tab)

    return moyenne

print(moyenne_temps_find_community_from_person(network, 1000))

0.013989999999999999


En comparant les fonctions find_community_by_decreasing_popularity et find_community_from_person, on s'aperçoit que la première met en moyenne 0.12 secondes à s'exécuter contre 0.014 pour la deuxième. L'écart est assez énorme, la première est environ 10 fois plus longue à s'exécuter que la deuxième. Cela s'explique par le fait qu'elle appelle dans sa définition la fonction order_by_decreasing_popularity, qui est une fonction de complexité O(n²) + O(n), car elle contient une boucle et une boucle imbriquée. Ajoutons à cela que la fonction contient une boucle, donc en plus une opération de complexité O(n).